Software/lib/GNSE_HAL/GNSE_flash.c

```diff
@@ -0,0 +1,225 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file GNSE_flash.c
+ *
+ * @copyright Copyright (c) 2021 The Things Industries B.V.
+ *
+ */
+
+#include "GNSE_flash.h"
+#include "GNSE_bsp.h"
+
+MxChip GNSE_Flash;
+spiffs GNSE_Flash_SPIFFS;
+
+/**
+ * Buffers needed for SPIFFS, based on lib author requirements
+ * @warning: Avoid changing these paramteres unless you are familiar with the change effect
+ */
+#define GNSE_FLASH_PAGE_SIZE 256U
+#define GNSE_FLASH_FS_FD 4U
+#define GNSE_FLASH_FS_FD_SIZE 32U
+#define GNSE_FLASH_FS_MIN_WORK_BUFFER 2U
+
+static uint8_t spiffs_work_buf[GNSE_FLASH_PAGE_SIZE * GNSE_FLASH_FS_MIN_WORK_BUFFER];
+static uint8_t spiffs_fds[GNSE_FLASH_FS_FD_SIZE * GNSE_FLASH_FS_FD];
+static uint8_t spiffs_cache_buf[(GNSE_FLASH_PAGE_SIZE + GNSE_FLASH_FS_FD_SIZE) * GNSE_FLASH_FS_FD];
+
+/**
+  * @brief Initialize hardware of the external SPI flash
+  * @param none
+  * @return FLASH_op_result_t, see enum for more information
+  */
+FLASH_op_result_t GNSE_Flash_Init(void)
+{
+  GNSE_BSP_LS_Init(LOAD_SWITCH_FLASH);
+  GNSE_BSP_LS_On(LOAD_SWITCH_FLASH);
+  HAL_Delay(LOAD_SWITCH_FLASH_DELAY_MS);
+  if (MX25R16_Init(&GNSE_Flash) != MXST_SUCCESS)
+  {
+    return FLASH_OP_FAIL;
+  }
+  return FLASH_OP_SUCCESS;
+}
+
+/**
+  * @brief Deinitialize hardware of the external SPI flash
+  * @param none
+  * @return FLASH_op_result_t, see enum for more information
+  */
+FLASH_op_result_t GNSE_Flash_DeInit(void)
+{
+  GNSE_BSP_LS_Off(LOAD_SWITCH_FLASH);
+  GNSE_BSP_LS_DeInit(LOAD_SWITCH_FLASH);
+  GNSE_BSP_Flash_SPI_DeInit();
+  return FLASH_OP_SUCCESS;
+}
+
+/**
+ * @brief Reads a number of bytes from external flash
+ * This function abstracts MxREAD (found as part of MX25R16 APIs in nor_cmd.c)
+ *
+ * @param addr for MX25R16 SPI flash the address range is from 0x000000 to 0x1FFFFF => 16 Mb
+ * @param byte_count
+ * @param target_buffer the buffer that will be used to store the read data
+ * @return FLASH_op_result_t, see enum for more information
+ */
+FLASH_op_result_t GNSE_Flash_Read(uint32_t addr, uint32_t byte_count, uint8_t *target_buffer)
+{
+  FLASH_op_result_t status = FLASH_OP_SUCCESS;
+  if (MxIsFlashBusy(&GNSE_Flash) == MXST_DEVICE_READY)
+  {
+    if (GNSE_Flash.AppGrp._Read(&GNSE_Flash, addr, byte_count, target_buffer) != MXST_SUCCESS)
+    {
+      status = FLASH_OP_FAIL;
+    }
+  }
+  else
+  {
+    status = FLASH_OP_FAIL;
+  }
+
+  return status;
+}
+
+/**
+ * @brief Write a number of bytes to external flash
+ * This function abstracts MxPP (found as part of MX25R16 APIs in nor_cmd.c)
+ *
+ * @param addr for MX25R16 SPI flash the address range is from 0x000000 to 0x1FFFFF => 16 Mb
+ * @param byte_count
```

Comment:
You can leave the description empty, but I'll leave this comment just in case you forgot it

Reply:
Yeah if the parameter is descriptive on it's own, it can be left empty

Reply:
Great